File: scripts/nsd_cleanup_zones.pl
#!/usr/bin/perl

use strict;
use warnings;
use Net::DNS;

# Metazone name. Leave default unless you use another name for the metazone.
my $metazone = 'meta.meta';  

# Full Path to nsd-control 
my $nsdcontrol = '/usr/local/sbin/nsd-control';

# This Root NS Unicast (secondary) IP (where NSD listens to for AXFR requests)
my $rootns = '10.1.1.211'; 

# TMP Folder where nsd_superslave writes incoming zones (without trailing slash)
my $depot = '/data/tmp/zones';


my $res = Net::DNS::Resolver->new;

# Perform a zone transfer from the local slave server
$res->nameservers($rootns);

my @zone = $res->axfr($metazone);

foreach my $rr (@zone) {
    # Skip records we know can't be meant for us
    next unless $rr->type eq 'TXT';
    next unless $rr->name =~ /\.${metazone}$/;

    # Strip meta zone name from origin
    my $name = $rr->name;
    $name =~ s/\.${metazone}$//;

    # Perform sanity check on name
    # Skip if zone $name doesn't exist on this slave.

    # Remove zone from slave server:
    #   if BIND:
    #     issue [ rndc delzone ] 
    #   if NSD:
    #     issue [ nsd-control delzone ]
    #   if PowerDNS slave:
    #    delete from back-end database
    #   etc.
    
    # Determine if zone file on disk; if so, move it into
    #    backup area.
     
    if (defined($name)) {
        my $command = "${nsdcontrol} delzone ${name} ";
        system($command);
        my $command2 = "rm -rf ". $depot ."/". $name ;
        system($command2);

		if ($metazone ne $name){
        	# clean stub zone from BIND
        	system("/usr/local/bin/bind_del_stub.sh ".$name );
		}

    }


    print $name, "\n";
}

File: scripts/nsd_cleanup_zones.php
#!/usr/bin/php
<?php
/*-----------------------------------------------------------------------------
* Domain Registry Control Panel                                               *
*                                                                             *
* Main Author: Vaggelis Koutroumpas vaggelis@koutroumpas.gr (c)2014 for AWMN  *
* Credits: see CREDITS file                                                   *
*                                                                             *
* This program is free software: you can redistribute it and/or modify        *
* it under the terms of the GNU General Public License as published by        * 
* the Free Software Foundation, either version 3 of the License, or           *
* (at your option) any later version.                                         *
*                                                                             *
* This program is distributed in the hope that it will be useful,             *
* but WITHOUT ANY WARRANTY; without even the implied warranty of              *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                *
* GNU General Public License for more details.                                *
*                                                                             *
* You should have received a copy of the GNU General Public License           *
* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
*                                                                             *
*-----------------------------------------------------------------------------*/

//CONFIGURATION
$CONF['unicast_ip'] = '10.1.1.214';
$CONF['metazone']   = 'meta.meta'; // leave this as is
$CONF['nsdcontrol'] = '/usr/local/sbin/nsd-control';
$CONF['zones_path'] = '/data/tmp/zones';
$CONF['deleted_zones'] = '/data/tmp/deleted_zones.txt';
$CONF['debug'] = false;
//END OF CONFIGURATION


$deleted_zones = file($CONF['deleted_zones']);
$deleted_zones = array_map('trim', $deleted_zones);

require_once 'Net/DNS.php';

$resolver = new Net_DNS_Resolver();
$resolver->debug = $CONF['debug'];
$resolver->nameservers = array($CONF['unicast_ip']);
$response = $resolver->axfr($CONF['metazone']);

if (count($response)) {
	foreach ($response as $rr) {
  		if ($rr->type == 'TXT'){
      		$record = str_replace(".".$CONF['metazone'], "", $rr->name);
      		if ($record && !in_array($record, $deleted_zones)){
                exec ($CONF['nsdcontrol'] . " delzone " . $record, $delzone);
				if ($delzone[0] == 'ok'){
					if (file_exists($CONF['zones_path'] . "/" . $record)){
						exec ("rm -rf " . $CONF['zones_path'] . "/" . $record);
					}
					file_put_contents($CONF['deleted_zones'], $record."\n", FILE_APPEND);
					echo "Deleted zone " . $record . " from NSD.\n";
				}elseif (strstr($delzone[0], "not present")){
					file_put_contents($CONF['deleted_zones'], $record."\n", FILE_APPEND);
					echo "Zone " . $record . " was already deleted from NSD. Updated DB.\n";										
				}else{
					echo "Could not delete zone " . $record . " from NSD\n";
				}			
        	}else{
				if ($CONF['debug']){
					echo "Zone " . $record. " is already deleted\n";
				}
        	}
    	}
	}
}else{
	echo "Could not AXFR zone " . $CONF['metazone'] . " from " . $CONF['unicast_ip'] . "\n";
}


?>